core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java

```diff
@@ -587,10 +620,33 @@ public DriverSplitRunner createDriverRunner(@Nullable ScheduledSplit partitioned
             return new DriverSplitRunner(this, driverContext, partitionedSplit);
         }
 
+        /**
+         * @return the created {@link Driver}, or <code>null</code> if the driver factory is already closed because the task is terminating
+         */
+        @Nullable
         public Driver createDriver(DriverContext driverContext, @Nullable ScheduledSplit partitionedSplit)
         {
-            Driver driver = driverFactory.createDriver(driverContext);
+            // Attempt to increment the driver count eagerly, but skip driver creation if the task is already terminating or done
+            if (!driverAndTaskTerminationTracker.tryCreateNewDriver()) {
+                return null;
+            }
+            Driver driver;
+            try {
+                driver = driverFactory.createDriver(driverContext);
+            }
+            catch (Exception e) {
+                // run the destroyed listener since the live driver count was incremented, but driver creation failed
+                driverAndTaskTerminationTracker.driverDestroyed();
+                // driverFactory is already closed, ignore the exception and return null
+                if (driverFactory.isNoMoreDrivers()) {
+                    return null;
+                }
+                // this exception is unexpected if driverFactory has not been closed, so rethrow it
+                throw e;
+            }
 
+            // register driver destroyed listener to detect when termination completes
+            driver.getDestroyedFuture().addListener(driverAndTaskTerminationTracker, directExecutor());
```

Comment:
I'm not saying this is bad, but make sure you actually want direct executor here... depends on the on the context where destroyed future is set.

Reply:
For this particular case, I'm fairly confident that we want to use a direct executor. The action is short, only needs to perform an atomic update and maybe flip the state machine if termination has finished, and by doing it this way we know that we won't have a `RejectedExecutionException` causing tasks to get stuck and never finish terminating.